xmppserver/src/main/java/org/jivesoftware/openfire/http/HttpSession.java

```diff
@@ -503,26 +511,24 @@ public void forwardRequest(HttpBindBody body, AsyncContext context)
              * deliverable on the new connection. This is under the assumption that a connection has been dropped,
              * and re-requested before jetty has realised.
              */
-            // TODO is there overlap with the retransmission logic above? Can this be simplified?
-            for (HttpConnection queuedConnection : connectionQueue) {
+            final Iterator<HttpConnection> iter = connectionQueue.iterator();
+            while (iter.hasNext()) {
+                final HttpConnection queuedConnection = iter.next();
                 if (queuedConnection.getRequestId() == rid) {
                     Log.debug("Found previous connection in queue with rid {}", rid);
-                    if(queuedConnection.isClosed()) {
-                        Log.debug("It's closed - copying deliverables");
 
-                        Delivered deliverable = retrieveDeliverable(rid);
-                        if (deliverable == null) {
-                            Log.warn("In session {} deliverable unavailable for {}", streamid, rid);
-                            throw new HttpBindException("Unexpected RID error.", BoshBindingError.itemNotFound);
-                        }
-                        connection.deliverBody(asBodyText(deliverable.deliverables), true);
-                    } else {
-                        Log.debug("For session {} queued connection is still open - calling close() on the old connection (as the new connection will replace it).", streamid);
-
-                        // TODO implement section 14.3 of XEP 0124 instead of this!
-                        deliver(queuedConnection, Collections.singletonList(new Deliverable("")), true);
-                        connection.close();
-                    }
+                    // Note that the old connection is removed here, but the new connection will not be added back before the mutex is released. This leaves
+                    // room for another thread to try and consume a connection before this connection has been restored. This should be safe, as the consumer
+                    // should not be allowed to consume a connection with a RID that, compared to the previously consumed RID, leaves a 'gap'. In that case,
+                    // the to-be-delivered data is expected to be queued, which will be picked up as soon as this connection (for which the RID 'fills the gap',
+                    // is being processed by org.jivesoftware.openfire.http.HttpSession.processConnection.
+                    iter.remove();
+
+                    assert !queuedConnection.isClosed();
```

Comment:
Is assert the right thing to do here? Perhaps a log statement with some tidy up?

Reply:
Recently, I have started leaning towards failing fast(er) than what we used to do. The connection at this stage is really supposed to be closed. If it's not, that is a bug in code. My current way of thinking is that I'd rather assert this state (which has the benefit of documenting the intent of this code) than write additional code that kicks in only when the preceding code has a bug. That feels like putting on bandages, just in case an injury will occur.

Over time, it's hard to distinguish between the "just in case" code, and the code that is actually supposed to implement the desired functionality. I'm hoping that with using assertions, this will be prevented.